Crash when "terraform" block has syntax error

When running terraform init I receive runtime error: invalid memory address or nil pointer dereference.

I was able to successfully update some of our modules. But in our, what we call stacks, I am unable to run init. We usually wrap the terraform call in terragrunt as it provides some higher level functions (such as find_in_parent_folders() and path_relative_from_include() that allows us to determine the region and profile based off directory structure). However for this test I removed all that and hard coded the account and region. Still results in a crash. 

### Terraform Version
```
› terraform -v
Terraform v0.12.0-beta2
```

### Terraform Configuration Files

```hcl
/*
  This variable is auto-generated via terragrunt based on the relative directory location of
  terraform.tfvars to the include location (path_relative_to_include())

  Our directory structure is account-name/aws-region (ie: app-nprod/us-west-2) and a matching
  profile must exist in your awscli config/credentials.
*/
locals {
  // account and region are interpolated from the relative path
  //account   = "${element(split("/", var.path), 0)}"
  //region    = "${element(split("/", var.path), 1)}"

  // environment is interpolated from the relative path
  //environment = "${element(split("/", var.path), 0)}"

  /* Default tags (local so you can't over-ride) */
  tags {
    //product    = "${var.product}"
    //env        = "${local.environment}"
    built-with = "terraform"
  }
}

variable "path" { description = "Relative path to root that specifies environment and region" }

variable "product" { default = "routes" }

/* additional aliases that distribution is accessed through */
variable "aliases" { default = [] }

variable "tags" { default =  {} }

/*
  Setup the AWS provider
*/
provider aws {
//  profile = "${local.account}"
//  region =  "${local.region}"
  profile = "test"
  region  = "us-west-2"
}

provider aws {
  alias  = "us-east-1"
  region = "us-east-1"

  //profile = "${local.account}"
  profile = "test"
}

terraform { backend "s3" {} }

/*
  Setup remote state data sources
*/
data terraform_remote_state "security" {
  backend = "s3"
  config {
    bucket  = "com-dat-tf"
    region  = "us-west-2"
    key     = "security/test/global/terraform.tfstate"
  }
}

data terraform_remote_state "networking" {
  backend = "s3"
  config {
    bucket  = "com-dat-tf"
    region  = "us-west-2"
    key     = "networking/test/us-west-2/terraform.tfstate"
  }
}

data "terraform_remote_state" "shared" {
  backend = "s3"
  config {
    bucket = "com-dat-tf"
    region = "us-west-2"
    key    = "shared/test/us-west-2/terraform.tfstate"
  }
}

/* cloudfront distribution for SPA */
data aws_acm_certificate "certificate" {
  provider = "aws.us-east-1"

  /* look for wildcard cert that matches our public route53 zone */
  domain = "${format("*.%s", replace(replace(format("%s.dat.com", local.environment), "/^prod\\./", ""), "/\\.$/", ""))}"
}

module "angular-routes" {
  source = "git::ssh://git@bitbucket.org/dat/tf-aws-cloudfront.git?ref=1.8.0"

  name = "${format("%s.%s.dat.com", var.product, local.environment)}"

  /* ACM certificate for environment */
  acm_certificate_arn = "${data.aws_acm_certificate.certificate.arn}"


  aliases = [ "${var.aliases}" ]

  /* provider inheritance https://www.terraform.io/docs/modules/usage.html#providers-within-modules */
  providers {
    "aws.us-east-1" = "aws.us-east-1"
  }

  tags = "${merge(var.tags, local.tags)}"
}

/* eventually this will be incorparated into its own module */
resource aws_route53_record "cname" {
  zone_id = "${data.terraform_remote_state.networking.route53_public_zone_id[0]}"

  name = "${var.product}"

  records = [ "${module.angular-routes.cf_distribution_domain_name}" ]
  type    = "CNAME"
  ttl     = 300
}
```

### Debug Output

```
2019/05/13 15:46:13 [INFO] Terraform version: 0.12.0 beta2
2019/05/13 15:46:13 [INFO] Go runtime version: go1.11.5
2019/05/13 15:46:13 [INFO] CLI args: []string{"/Users/matthewce/bin/terraform", "init"}
2019/05/13 15:46:13 [DEBUG] Attempting to open CLI config file: /Users/matthewce/.terraformrc
2019/05/13 15:46:13 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2019/05/13 15:46:13 [INFO] CLI command args: []string{"init"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x13fa5f0]

goroutine 1 [running]:
github.com/hashicorp/hcl2/hcl/hclsyntax.(*Body).PartialContent(0x0, 0x3855460, 0xc0004ce120, 0x2a16de0, 0xc0004644d0, 0x0, 0x0, 0x0)
        /opt/gopath/pkg/mod/github.com/hashicorp/hcl2@v0.0.0-20190416162332-2c5a4b7d729a/hcl/hclsyntax/structure.go:135 +0x80
github.com/hashicorp/terraform/configs.sniffCoreVersionRequirements(0x2a16de0, 0xc000464420, 0x9, 0x2a16de0, 0xc000464420, 0xc0002ba830, 0x2, 0x2)
        /opt/gopath/src/github.com/hashicorp/terraform/configs/parser_config.go:156 +0x121
github.com/hashicorp/terraform/configs.(*Parser).loadConfigFile(0xc0004a8000, 0xc0002be4a0, 0x9, 0x0, 0x0, 0xc0002b7cb0, 0xc, 0x8)
        /opt/gopath/src/github.com/hashicorp/terraform/configs/parser_config.go:42 +0xdc
github.com/hashicorp/terraform/configs.(*Parser).LoadConfigFile(0xc0004a8000, 0xc0002be4a0, 0x9, 0x0, 0x0, 0x7, 0xc0002be422)
        /opt/gopath/src/github.com/hashicorp/terraform/configs/parser_config.go:22 +0x44
github.com/hashicorp/terraform/configs.(*Parser).loadFiles(0xc0004a8000, 0xc0002ca880, 0x3, 0x4, 0x0, 0x4, 0x0, 0x0, 0x0, 0x0, ...)
        /opt/gopath/src/github.com/hashicorp/terraform/configs/parser_config_dir.go:76 +0x349
github.com/hashicorp/terraform/configs.(*Parser).LoadConfigDir(0xc0004a8000, 0x26c734d, 0x1, 0x0, 0x1, 0xc0003e7430, 0x1757710)
        /opt/gopath/src/github.com/hashicorp/terraform/configs/parser_config_dir.go:36 +0x124
github.com/hashicorp/terraform/command.(*Meta).loadSingleModule(0xc0004776c0, 0x26c734d, 0x1, 0x0, 0x0, 0x0, 0xe)
        /opt/gopath/src/github.com/hashicorp/terraform/command/meta_config.go:113 +0x158
github.com/hashicorp/terraform/command.(*InitCommand).Run(0xc0004776c0, 0xc00000a070, 0x0, 0x0, 0xc0002ba100)
        /opt/gopath/src/github.com/hashicorp/terraform/command/init.go:182 +0x9fc
github.com/mitchellh/cli.(*CLI).Run(0xc000458280, 0xc000458280, 0xc0003e7d20, 0x1)
        /opt/gopath/pkg/mod/github.com/mitchellh/cli@v1.0.0/cli.go:255 +0x207
main.wrappedMain(0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:223 +0xb37
main.realMain(0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:100 +0xb7
main.main()
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:36 +0x2e
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x13fa5f0]

goroutine 1 [running]:
github.com/hashicorp/hcl2/hcl/hclsyntax.(*Body).PartialContent(0x0, 0x3855460, 0xc0004ce120, 0x2a16de0, 0xc0004644d0, 0x0, 0x0, 0x0)
        /opt/gopath/pkg/mod/github.com/hashicorp/hcl2@v0.0.0-20190416162332-2c5a4b7d729a/hcl/hclsyntax/structure.go:135 +0x80
github.com/hashicorp/terraform/configs.sniffCoreVersionRequirements(0x2a16de0, 0xc000464420, 0x9, 0x2a16de0, 0xc000464420, 0xc0002ba830, 0x2, 0x2)
        /opt/gopath/src/github.com/hashicorp/terraform/configs/parser_config.go:156 +0x121
github.com/hashicorp/terraform/configs.(*Parser).loadConfigFile(0xc0004a8000, 0xc0002be4a0, 0x9, 0x0, 0x0, 0xc0002b7cb0, 0xc, 0x8)
        /opt/gopath/src/github.com/hashicorp/terraform/configs/parser_config.go:42 +0xdc
github.com/hashicorp/terraform/configs.(*Parser).LoadConfigFile(0xc0004a8000, 0xc0002be4a0, 0x9, 0x0, 0x0, 0x7, 0xc0002be422)
        /opt/gopath/src/github.com/hashicorp/terraform/configs/parser_config.go:22 +0x44
github.com/hashicorp/terraform/configs.(*Parser).loadFiles(0xc0004a8000, 0xc0002ca880, 0x3, 0x4, 0x0, 0x4, 0x0, 0x0, 0x0, 0x0, ...)
        /opt/gopath/src/github.com/hashicorp/terraform/configs/parser_config_dir.go:76 +0x349
github.com/hashicorp/terraform/configs.(*Parser).LoadConfigDir(0xc0004a8000, 0x26c734d, 0x1, 0x0, 0x1, 0xc0003e7430, 0x1757710)
        /opt/gopath/src/github.com/hashicorp/terraform/configs/parser_config_dir.go:36 +0x124
github.com/hashicorp/terraform/command.(*Meta).loadSingleModule(0xc0004776c0, 0x26c734d, 0x1, 0x0, 0x0, 0x0, 0xe)
        /opt/gopath/src/github.com/hashicorp/terraform/command/meta_config.go:113 +0x158
github.com/hashicorp/terraform/command.(*InitCommand).Run(0xc0004776c0, 0xc00000a070, 0x0, 0x0, 0xc0002ba100)
        /opt/gopath/src/github.com/hashicorp/terraform/command/init.go:182 +0x9fc
github.com/mitchellh/cli.(*CLI).Run(0xc000458280, 0xc000458280, 0xc0003e7d20, 0x1)
        /opt/gopath/pkg/mod/github.com/mitchellh/cli@v1.0.0/cli.go:255 +0x207
main.wrappedMain(0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:223 +0xb37
main.realMain(0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:100 +0xb7
main.main()
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:36 +0x2e
```

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->

### Expected Behavior

Should not crash.

### Actual Behavior

Results in crash log. 

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`

```hcl
~/code/terraform/stacks/routes/test/us-west-2/ on git : master
› terraform -v
Terraform v0.12.0-rc1


~/code/terraform/stacks/routes/test/us-west-2/ on git : master
› terraform init
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x142ff1d]

goroutine 1 [running]:
github.com/hashicorp/hcl2/hcl/hclsyntax.(*Body).PartialContent(0x0, 0x3b13380, 0xc00048c540, 0x2bde6e0, 0xc0000e84d0, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/vendor/github.com/hashicorp/hcl2/hcl/hclsyntax/structure.go:135 +0x6d
github.com/hashicorp/terraform/configs.sniffCoreVersionRequirements(0x2bde6e0, 0xc0000e8420, 0x9, 0x2bde6e0, 0xc0000e8420, 0xc00007c850, 0x2, 0x2)
        /opt/gopath/src/github.com/hashicorp/terraform/configs/parser_config.go:156 +0x121
github.com/hashicorp/terraform/configs.(*Parser).loadConfigFile(0xc0002d0320, 0xc000040390, 0x9, 0x0, 0x2, 0x0, 0x0, 0x7)
        /opt/gopath/src/github.com/hashicorp/terraform/configs/parser_config.go:42 +0xdc
github.com/hashicorp/terraform/configs.(*Parser).LoadConfigFile(...)
        /opt/gopath/src/github.com/hashicorp/terraform/configs/parser_config.go:22
github.com/hashicorp/terraform/configs.(*Parser).loadFiles(0xc0002d0320, 0xc00004c1c0, 0x3, 0x4, 0x0, 0x4, 0x0, 0x0, 0x0, 0x0, ...)
        /opt/gopath/src/github.com/hashicorp/terraform/configs/parser_config_dir.go:76 +0x343
github.com/hashicorp/terraform/configs.(*Parser).LoadConfigDir(0xc0002d0320, 0x27b048d, 0x1, 0x0, 0x1, 0x3, 0x4)
        /opt/gopath/src/github.com/hashicorp/terraform/configs/parser_config_dir.go:36 +0x121
github.com/hashicorp/terraform/command.(*Meta).loadSingleModule(0xc0004d1380, 0x27b048d, 0x1, 0x0, 0x0, 0x0, 0xe)
        /opt/gopath/src/github.com/hashicorp/terraform/command/meta_config.go:113 +0x136
github.com/hashicorp/terraform/command.(*InitCommand).Run(0xc0004d1380, 0xc0000be010, 0x0, 0x0, 0xc0000c19e0)
        /opt/gopath/src/github.com/hashicorp/terraform/command/init.go:182 +0xad4
github.com/mitchellh/cli.(*CLI).Run(0xc000562500, 0xc000562500, 0xc000531d90, 0x1)
        /opt/gopath/src/github.com/hashicorp/terraform/vendor/github.com/mitchellh/cli/cli.go:255 +0x1f1
main.wrappedMain(0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:223 +0xaff
main.realMain(0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:100 +0xb4
main.main()
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:36 +0x3b



!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

~/code/terraform/stacks/routes/test/us-west-2/ on git : master
```

--- a/command/init_test.go
+++ b/command/init_test.go
@@ -1310,3 +1310,33 @@ func TestInit_012UpgradeNeededInAutomation(t *testing.T) {
 		t.Errorf("looks like we incorrectly gave an upgrade command to run:\n%s", output)
 	}
 }
+
+func TestInit_syntaxErrorVersionSniff(t *testing.T) {
+	// Create a temporary working directory that is empty
+	td := tempDir(t)
+	copy.CopyDir(testFixturePath("init-sniff-version-error"), td)
+	defer os.RemoveAll(td)
+	defer testChdir(t, td)()
+
+	ui := new(cli.MockUi)
+	c := &InitCommand{
+		Meta: Meta{
+			testingOverrides: metaOverridesForProvider(testProvider()),
+			Ui:               ui,
+		},
+	}
+
+	args := []string{}
+	if code := c.Run(args); code != 0 {
+		t.Fatalf("bad: \n%s", ui.ErrorWriter.String())
+	}
+
+	// Check output.
+	// Currently, this lands in the "upgrade may be needed" codepath, because
+	// the intentional syntax error in our test fixture is something that
+	// "terraform 0.12upgrade" could fix.
+	output := ui.OutputWriter.String()
+	if got, want := output, "Terraform has initialized, but configuration upgrades may be needed"; !strings.Contains(got, want) {
+		t.Fatalf("wrong output\ngot:\n%s\n\nwant: message containing %q", got, want)
+	}
+}

--- a/command/test-fixtures/init-sniff-version-error/init-sniff-version-error.tf
+++ b/command/test-fixtures/init-sniff-version-error/init-sniff-version-error.tf
@@ -1,0 +1,9 @@
+# The following is invalid because we don't permit multiple nested blocks
+# all one one line. Instead, we require the backend block to be on a line
+# of its own.
+# The purpose of this test case is to see that HCL still produces a valid-enough
+# AST that we can try to sniff in this block for a terraform_version argument
+# without crashing, since we do that during init to try to give a better
+# error message if we detect that the configuration is for a newer Terraform
+# version.
+terraform { backend "local" {} }

--- a/go.mod
+++ b/go.mod
@@ -60,7 +60,7 @@ require (
 	github.com/hashicorp/go-version v1.1.0
 	github.com/hashicorp/golang-lru v0.5.0 // indirect
 	github.com/hashicorp/hcl v0.0.0-20170504190234-a4b07c25de5f
-	github.com/hashicorp/hcl2 v0.0.0-20190503210054-6e4ec17113ca
+	github.com/hashicorp/hcl2 v0.0.0-20190514214226-6a61d80ae3d0
 	github.com/hashicorp/hil v0.0.0-20190212112733-ab17b08d6590
 	github.com/hashicorp/logutils v1.0.0
 	github.com/hashicorp/memberlist v0.1.0 // indirect

